go-controller/pkg/ovn/pods.go

```diff
@@ -170,42 +169,67 @@ func getRoutesGatewayIP(pod *kapi.Pod, subnet *net.IPNet, hybridOverlayExternalG
 			break
 		}
 	}
-	gatewayIPnet := util.GetNodeGatewayIfAddr(subnet)
-	var gatewayIP net.IP
-	routes := make([]util.PodRoute, 0)
-	if otherDefaultRoute || len(hybridOverlayExternalGW) > 0 {
-		for _, clusterSubnet := range config.Default.ClusterSubnets {
-			var route util.PodRoute
-			route.Dest = clusterSubnet.CIDR
-			route.NextHop = gatewayIPnet.IP
-			routes = append(routes, route)
+	// DUALSTACK FIXME: hybridOverlayExternalGW is not Dualstack
+	hybridOverlayExternalGWFromAnnt := net.IP{}
+	if config.HybridOverlay.Enabled {
+		hybridOverlayExternalGWFromAnnt, err = oc.getHybridOverlayExternalGwAnnotation(pod.Namespace)
+		if err != nil {
+			return err
 		}
-		for _, serviceSubnet := range config.Kubernetes.ServiceCIDRs {
-			var route util.PodRoute
-			route.Dest = serviceSubnet
-			route.NextHop = gatewayIPnet.IP
-			routes = append(routes, route)
+	}
+
+	for _, podIfAddr := range podAnnotation.IPs {
+		isIPv6 := utilnet.IsIPv6CIDR(podIfAddr)
+		nodeSubnet, err := util.MatchIPFamily(isIPv6, nodeSubnets)
+		if err != nil {
+			return err
 		}
-		if len(hybridOverlayExternalGW) > 0 {
-			gatewayIP = util.GetNodeHybridOverlayIfAddr(subnet).IP
+
+		// DUALSTACK FIXME: hybridOverlayExternalGW is not Dualstack
+		hybridOverlayExternalGW := net.IP{}
+		if utilnet.IsIPv6(hybridOverlayExternalGWFromAnnt) == isIPv6 {
+			hybridOverlayExternalGW = hybridOverlayExternalGWFromAnnt
 		}
-	} else {
-		gatewayIP = gatewayIPnet.IP
-	}
 
-	if gatewayIP != nil && len(config.HybridOverlay.ClusterSubnets) > 0 {
-		// Add a route for each hybrid overlay subnet via the hybrid
-		// overlay port on the pod's logical switch.
-		second := util.NextIP(gatewayIP)
-		thirdIP := util.NextIP(second)
-		for _, subnet := range config.HybridOverlay.ClusterSubnets {
-			routes = append(routes, util.PodRoute{
-				Dest:    subnet.CIDR,
-				NextHop: thirdIP,
-			})
+		gatewayIPnet := util.GetNodeGatewayIfAddr(nodeSubnet)
+		var gatewayIP net.IP
+		if otherDefaultRoute || len(hybridOverlayExternalGW) > 0 {
+			for _, clusterSubnet := range config.Default.ClusterSubnets {
+				var route util.PodRoute
+				route.Dest = clusterSubnet.CIDR
+				route.NextHop = gatewayIPnet.IP
+				podAnnotation.Routes = append(podAnnotation.Routes, route)
+			}
+			for _, serviceSubnet := range config.Kubernetes.ServiceCIDRs {
+				var route util.PodRoute
+				route.Dest = serviceSubnet
+				route.NextHop = gatewayIPnet.IP
+				podAnnotation.Routes = append(podAnnotation.Routes, route)
+			}
+			if len(hybridOverlayExternalGW) > 0 {
+				gatewayIP = util.GetNodeHybridOverlayIfAddr(nodeSubnet).IP
+			}
+		} else {
+			gatewayIP = gatewayIPnet.IP
+		}
+
+		if gatewayIP != nil && len(config.HybridOverlay.ClusterSubnets) > 0 {
+			// Add a route for each hybrid overlay subnet via the hybrid
+			// overlay port on the pod's logical switch.
+			second := util.NextIP(gatewayIP)
+			thirdIP := util.NextIP(second)
```

Comment:
more bad existing code to fix.

    nextHop := util.GetNodeHybridOverlayIfAddr(nodeSubnet)

(and then `NextHop: nextHop.IP` below)

Actually, we shouldn't be checking `gatewayIP != nil` for this `if` block either. If there are hybrid overlay cluster subnets, then we _must_ route traffic for them through the hybrid overlay gateway, regardless of whether `otherDefaultRoute || hybridOverlayExternalGW != nil` was true above.